---
title: "Filtered_fly_genome"
author: "Isabel Birds"
date: "7 December 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(rtracklayer)
library(dplyr)
library(seqinr)
```

One CDS per gene to start.
gtf from http://ftp.flybase.net/releases/FB2020_04/dmel_r6.35/gtf/dmel-all-r6.35.gtf.gz 

```{r load}

#My curated files
fly_fasta <- read.fasta("../2_Processed_data/Dmel_genomes/Dmel.fasta",
                        seqtype = "DNA",forceDNAtolower = FALSE)
fly_gff <- readGFF("../2_Processed_data/Dmel_genomes/Dmel.gff3")

#All genes info
fly_base <- readGFF("../1_Raw_data/dmel-all-r6.35.gtf") %>% filter(type=="CDS")
```

Open GTF, pull out all CDS IDs.
Filter to one CDS per gene.

```{r ID_CDS} 

#Filter all genes to my curated transcripts.
# Join gene ID to CDSs, find lenght of each CDS
fly_base <- fly_base[fly_base$transcript_symbol %in% fly_gff$seqid,] %>% 
  select(gene_id,transcript_symbol) %>% unique()
fly_base <- left_join(fly_gff,fly_base,by=c("seqid"="transcript_symbol")) %>% filter(type=="CDS")
fly_base$length <- fly_base$end - fly_base$start

#Select longest CDS, if no longest, random.
fly_base <- fly_base %>% 
  arrange(desc(length)) %>%
  group_by(gene_id) %>%
  slice(1)

```

Filter fasta

```{r fasta}

filtered_fasta <- fly_fasta[names(fly_fasta) %in% fly_base$seqid]

write.fasta(filtered_fasta,names=names(filtered_fasta),
            file.out ="../2_Processed_data/Dmel_genomes/Dmel_filtered.fasta")

```

Filter GFF

Some URL Encoding not saving correctly...

Having to replace:
%27 '
%29 )
%28 (
%5b [
%5d ]

```{r GFF}

filtered_fly_gff <- fly_gff[fly_gff$seqid %in% fly_base$seqid,]

export.gff3(filtered_fly_gff,con = "../2_Processed_data/Dmel_genomes/Dmel_filtered.gff3",)

```


